### PR TITLE
[wasm][debugger] Improvement of memory consumption while Evaluating Expressions

### PR DESCRIPTION
**Pull request by thaystg**
Removing the usage of CSharpCompilation which was increasing the memory consumption and using CSharpScript.

### CODE REVIEW

File: src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs [hunk 1]
```diff
@@ -222,19 +213,37 @@ private object ConvertJSToCSharpType(JToken variable)
...
+                        return $"\"{value?.Value<string>()}\"";
                     case "number":
                         return value?.Value<double>();
                     case "boolean":
-                        return value?.Value<bool>();
+                        return value?.Value<string>().ToLower();
                     case "object":
-                        return variable;
+                    {
+                        if (subType == "null")
```
Comment (lewing) [inline src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs:226]: is subtype ever anything but null here?

File: src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs [hunk 2]
```diff
@@ -248,14 +257,11 @@ private string GetTypeFullName(JToken variable)
                 switch (type)
                 {
                     case "object":
-                        {
-                            if (subType == "null")
```
Comment (lewing) [inline src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs:252]: looks like subType is unused now

File: src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs [hunk 3]
```diff
@@ -266,13 +272,8 @@ private string GetTypeFullName(JToken variable)
         private static SyntaxNode GetExpressionFromSyntaxTree(SyntaxTree syntaxTree)
```
Comment (lewing) [inline src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs:272]: replace calls to this with GetCompilationUnitRoot()?

File: src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs [hunk 4]
```diff
@@ -114,7 +117,7 @@ public SyntaxTree ReplaceVars(SyntaxTree syntaxTree, IEnumerable<JObject> ma_val
                     {
                         // Generate a random suffix
                         string suffix = Guid.NewGuid().ToString().Substring(0, 5);
-                        string prefix = iesStr.Trim().Replace(".", "_").Replace("(", "_").Replace(")", "_");
+                        string prefix = iesStr.Trim().Replace(".", "_").Replace("(", "_").Replace(")", "_").Replace(", ", "_").Replace("\"", "_").Replace("'", "_");
```
Comment (lewing) [inline src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs:120]: Is this complete? should it be a Regex.Replace?

File: src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs (hunk 1 above)
Comment (thaystg) [reply src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs:226]: Could also be an array, but in case of being an array it will have an objectId, so it's correct to run the code outside the if.

Comment (lewing) [reply src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs:227]: so we don't need to pass the subtype through in that case?

Comment (thaystg) [reply src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs]: You are right, I've changed the code now :)
Thanks

File: src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs [hunk 5]
```diff
@@ -97,7 +101,7 @@ public SyntaxTree ReplaceVars(SyntaxTree syntaxTree, IEnumerable<JObject> ma_val
                     {
                         // Generate a random suffix
                         string suffix = Guid.NewGuid().ToString().Substring(0, 5);
-                        string prefix = ma_str.Trim().Replace(".", "_");
+                        string prefix = Regex.Replace(ma_str, @"[^A-Za-z0-9_]", "_");
```
Comment (radical) [inline src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs:104]: should we have this regex cached?

Comment (thaystg) [reply src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs]: Done.

File: src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs [hunk 6]
```diff
@@ -222,59 +215,56 @@ private object ConvertJSToCSharpType(JToken variable)
                 JToken value = variable["value"];
                 string type = variable["type"].Value<string>();
                 string subType = variable["subtype"]?.Value<string>();
-
+                string objectId = variable["objectId"]?.Value<string>();
                 switch (type)
                 {
                     case "string":
-                        return value?.Value<string>();
+                        return $"\"{value?.Value<string>()}\"";
```
Comment (lewing) [inline src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs:222]: does this need additional escaping?

Comment (lewing) [inline src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs:222]: does this need additional escaping?

File: src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs [hunk 7]
```diff
@@ -32,6 +36,7 @@ private class FindVariableNMethodCall : CSharpSyntaxWalker
             private int visitCount;
             public bool hasMethodCalls;
             public bool hasElementAccesses;
+            internal string variableDefinition;
```
Comment (radical) [inline src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs:39]: Maybe use a `StringBuilder` since we are appending to it multiple times.

File: src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs [hunk 8]
```diff
@@ -409,48 +387,24 @@ public static object Evaluate()
...
-                syntaxTrees: new[] { syntaxTree },
-                references: references,
-                options: new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary));
-
-            SemanticModel semanticModel = compilation.GetSemanticModel(syntaxTree);
-            CodeAnalysis.TypeInfo TypeInfo = semanticModel.GetTypeInfo(expressionTree, cancellationToken: token);
-
-            using (var ms = new MemoryStream())
+            try {
+                var task = CSharpScript.EvaluateAsync(
```
Comment (radical) [inline src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs:395]: Why are we not `await`ing here?

File: src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs [hunk 9]
```diff
@@ -222,59 +215,56 @@ private object ConvertJSToCSharpType(JToken variable)
...
-                        {
-                            if (subType == "null")
-                                return variable["className"].Value<string>();
-                            else
-                                return "object";
-                        }
+                        return "object";
                     case "void":
                         return "object";
+                    case "boolean":
```
Comment (radical) [inline src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs:259]: add this case to the tests too

File: src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs [hunk 10]
```diff
@@ -222,59 +215,56 @@ private object ConvertJSToCSharpType(JToken variable)
                 JToken value = variable["value"];
                 string type = variable["type"].Value<string>();
                 string subType = variable["subtype"]?.Value<string>();
-
+                string objectId = variable["objectId"]?.Value<string>();
                 switch (type)
```
Comment (radical) [inline src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs:219]: Just checking, do we have tests that cover all these types?

File: src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs (hunk 9 above)
Comment (thaystg) [reply src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs:259]: This line is testing it:                    ("this.CallMethodWithParmBool(this.t)", TString("TRUE")),

File: src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs (hunk 6 above)
Comment (thaystg) [reply src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs]: Done.

File: src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs (hunk 10 above)
Comment (thaystg) [reply src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs:219]: Yes, I have to do these changes because the tests were failing, after the change to use CSharpScript.

File: src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs (hunk 8 above)
Comment (thaystg) [reply src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs]: Done.

File: src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs (hunk 7 above)
Comment (thaystg) [reply src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs]: Done.

File: src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs (hunk 6 above)
Comment (thaystg) [reply src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs]: Added a test for it also.

File: src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs (hunk 9 above)
Comment (radical) [reply src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs:263]: Ah, and this was being handled by `default: return value.GetType().FullName;` earlier? Why the additional case then?

Comment (thaystg) [reply src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs:263]: Because I did this change:
image

And this change was needed because otherwise when the variable was being declared and used it was creating like this: `var exampleVar4893742 = True;` and the first letter of the True/False with uppercase throws a compilation error.

Comment (radical) [reply src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs:263]: ok, that makes sense. IIUC, `ConvertJSToCSharpType` is being called once to get the value, and then `GetTypeFullName` also calls it. Instead, I think it would make sense for `ConvertJSToCSharpType` to return a `(string type, string value)`, since the type that we want to use in the final expression is tied closely to the "value" that we return.

And the method can be renamed to something like `ConvertJSToCSharpVariable`.

File: src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs [hunk 11]
```diff
@@ -32,6 +36,7 @@ private class FindVariableNMethodCall : CSharpSyntaxWalker
             private int visitCount;
             public bool hasMethodCalls;
             public bool hasElementAccesses;
+            internal StringBuilder variableDefinition = new StringBuilder();
```
Comment (radical) [inline src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs:39]: Reading the patch again, I understand that this is essentially all the locals that are being created for various parts of the expression. Given that, this can be renamed to `variableDefinitions`. and we could use a List<string> instead, so it wouldn't need contiguous space for the full string.

File: src/mono/wasm/debugger/DebuggerTestSuite/EvaluateOnCallFrameTests.cs [hunk 12]
```diff
@@ -542,6 +542,7 @@ async Task EvaluateOnCallFrameFail(string call_frame_id, params (string expressi
                     ("this.CallMethodWithParmBool(true)", TString("TRUE")),
                     ("this.CallMethodWithParmBool(false)", TString("FALSE")),
                     ("this.CallMethodWithParmString(\"concat\")", TString("str_const_concat")),
+                    ("this.CallMethodWithParmString(\"\\\"\\\"\")", TString("str_const_\"\"")),
```
Comment (lewing) [inline src/mono/wasm/debugger/DebuggerTestSuite/EvaluateOnCallFrameTests.cs:545]: I couldn't help myself
```suggestion csharp
                    ("this.CallMethodWithParmString(\"\\\"\\\"\")", TString("str_const_\"\"")),
                    ("this.CallMethodWithParmString(\"🛶\")", TString("str_const_🛶")),
                    ("this.CallMethodWithParmString(\"\\uD83D\\uDE80\")", TString("str_const_🛶")),
```
fails in the best possible way
```
[[this.CallMethodWithParmString("\uD83D\uDE80")] Value for json property named value didn't match.]

Expected: str_const_🛶
Actual:   str_const_🚀 
Expected: {
  "type": "string",
  "value": "str_const_🛶"
} 
Actual: {
  "type": "string",
  "value": "str_const_🚀",
  "description": "str_const_🚀"
}
CheckValue failed for this.CallMethodWithParmString("\uD83D\uDE80"). Expected: {
  "type": "string",
  "value": "str_const_🛶"
}, vs {
  "type": "string",
  "value": "str_const_🚀",
  "description": "str_const_🚀"
}
```

Comment (lewing) [reply src/mono/wasm/debugger/DebuggerTestSuite/EvaluateOnCallFrameTests.cs:545]: If you rebase on main and use
```
                    ("this.CallMethodWithParmString(\"🛶\")", TString("str_const_🛶")),
                    ("this.CallMethodWithParmString(\"\\uD83D\\uDEF6\")", TString("str_const_🛶")),
                    ("this.CallMethodWithParmString(\"🚀\")", TString("str_const_🚀")),
```
you should be fine

**Review by lewing — DISMISSED** [review]

File: src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs [hunk 13]
```diff
@@ -185,96 +190,73 @@ public SyntaxTree ReplaceVars(SyntaxTree syntaxTree, IEnumerable<JObject> ma_val
...
-                    case "void":
-                        return "object";
+                        valueRet = "Newtonsoft.Json.Linq.JObject.FromObject(new {"
+                                + $"type = \"object\","
+                                + $"description = \"object\","
+                                + $"className = \"object\","
+                                + $"subtype = \"null\""
+                                + "})";
+                        typeRet = "object";
+                        break;
```
Comment (radical) [inline src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs:252]: I'm not sure I understand these two.
1. why do we need to use JObjects in the script?
2. `case "object":` - would it be better to use `Newtonsoft.Json.Linq.JObject.Parse(..)` and pass the escaped string for `variable`?
3. `case "void":` - why is this needed? And what does the `variable` look like here? And do we have a test for this?

File: src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs [hunk 14]
```diff
@@ -409,48 +379,25 @@ public static object Evaluate()
...
-            using (var ms = new MemoryStream())
+            try {
+                var result = await CSharpScript.EvaluateAsync(
+                    string.Join("\n", findVarNMethodCall.variableDefinitions) + "\nreturn " + syntaxTree.ToString(),
+                    ScriptOptions.Default.WithReferences(
+                        typeof(object).Assembly,
+                        typeof(Enumerable).Assembly,
+                        typeof(JObject).Assembly
+                        ),
+                    cancellationToken: token);
```
Comment (radical) [inline src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs:394]: If we specifically catch `CompilationErrorException` here, can we surface the error to VS? or maybe just return the exception message like on line 400.

Comment (radical) [reply src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs:394]: Do I understand it correctly that if the expression itself throws an exception, then we return the message on line 400? Can we detect this specific case of the expression itself throwing, and return a different message?

Comment (radical) [reply src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs:394]: > If we specifically catch `CompilationErrorException` here, can we surface the error to VS? or maybe just return the exception message like on line 400.

IIUC, this would be thrown if our constructed script fails to compile, so maybe the message doesn't need to be shown to the user.

File: src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs (hunk 13 above)
Comment (thaystg) [reply src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs:252]: 1) Because that is how we return simple variable values, so when it's something that we really need to execute we also keep the same way of returning the value.
2) I think it's faster to not parse the string? once this is already parsed we don't need to parse it again?
3) Yes, the tests evaluating method that return void.

File: src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs [hunk 15]
```diff
@@ -409,48 +386,21 @@ public static object Evaluate()
...
-                MetadataReference.CreateFromFile(typeof(Enumerable).Assembly.Location)
-            };
-
-            CSharpCompilation compilation = CSharpCompilation.Create(
-                "compileAndRunTheExpression",
-                syntaxTrees: new[] { syntaxTree },
-                references: references,
-                options: new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary));
+            try {
+                var newScript = script.ContinueWith(
```
Comment (radical) [inline src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs:394]: Is re-using script like this faster, or more memory efficient? And `script` is not affected at all by `ContinueWith`?

File: src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs [hunk 16]
```diff
@@ -409,48 +386,21 @@ public static object Evaluate()
...
+                    string.Join("\n", findVarNMethodCall.variableDefinitions) + "\nreturn " + syntaxTree.ToString());
 
-            SemanticModel semanticModel = compilation.GetSemanticModel(syntaxTree);
-            CodeAnalysis.TypeInfo TypeInfo = semanticModel.GetTypeInfo(expressionTree, cancellationToken: token);
+                var state = await newScript.RunAsync(cancellationToken: token);
 
-            using (var ms = new MemoryStream())
+                return JObject.FromObject(ConvertCSharpToJSType(state.ReturnValue, state.ReturnValue.GetType()));
```
Comment (radical) [inline src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs:399]: Should this also be checking `state.Exception`?

File: src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs (hunk 15 above)
Comment (thaystg) [reply src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs:394]: Memory efficient.

File: src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs (hunk 16 above)
Comment (thaystg) [reply src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs:399]: https://github.com/dotnet/runtime/issues/61665 the follow up issue to work in the cases where we get an exception while evaluating the expression.

File: src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs (hunk 15 above)
Comment (thaystg) [reply src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs:394]: I don't think it's affected.
I did a test declaring the same variable name in two ContinueWith also changing the datatype and it works correctly.

```
var newScript = Class1.script.ContinueWith(
                    "System.String Val1 = \"one\";\nSystem.Double Val2 = 2;\nreturn $\"Some {Val1} Value {Val2} End\";");
var state = await newScript.RunAsync();
Console.WriteLine("result " + state.ReturnValue);

newScript = Class1.script.ContinueWith(
                    "System.Double Val1 = 3;\nSystem.Double Val2 = 3;\nreturn $\"Some {Val1} Value {Val2} End\";");
state = await newScript.RunAsync();
Console.WriteLine("result " + state.ReturnValue);
```


Result:
result Some one Value 2 End
result Some 3 Value 3 End

**Review by radical — APPROVED** [review]
LGTM! Thank you for patiently working with the feedback :)

File: src/mono/wasm/debugger/BrowserDebugProxy/BrowserDebugProxy.csproj [hunk 17]
```diff
@@ -7,6 +7,7 @@
   </PropertyGroup>
 
   <ItemGroup>
+    <PackageReference Include="Microsoft.CodeAnalysis.CSharp.Scripting" Version="3.7.0" />
```
Comment (radical) [inline src/mono/wasm/debugger/BrowserDebugProxy/BrowserDebugProxy.csproj:10]: For future PR, we should consider using the common version (more than one!) property for this, being used in other places in the repo.

**Review by lewing — APPROVED** [review]
